content/docs/reference/design-docs/editoast-errors/_index.en.md

```diff
@@ -0,0 +1,430 @@
+enum CreateError {
+    #[error("will be overridden, but still useful for development")]
+    #[view_error(internal, name = "Database")]
+    Db(#[from] DbError),
+
+    #[error(transparent)] // shown to the user
+    #[view_error(user)]
+    InvalidForm(#[from] FormError)
+}
+
+async fn create(...) -> Result<Json<Resource>, CreateError> {
+    todo!()
+}
+
+#[derive(Debug, thiserror::Error, ViewError)]
+enum UpdateError {
+    #[view_error(internal)]
+    Db(#[from] DbError),
+
+    #[view_error(internal)]
+    Redis(#[from] RedisError),
+
+    #[error(transparent)]
+    #[view_error(user)]
+    InvalidForm(#[from] FormError)
+}
+
+async fn update(...) -> Result<(), UpdateError> {
+    todo!()
+}
+````
+
+## Ease composability
+
+### Nesting `ViewError`s
+
+We composed errors by nesting them thanks to `thiserror`. However, to compose and reuse `EditoastErrors`, we need a special flag so that when we attempt to serialize the error, we return the serialization of the source error directly.
+
+```rust
+#[derive(Debug, thiserror::Error, ViewError)]
+#[error("no such infra: {id}")]
+#[view_error(context, status = 404)]
+struct InfraNotFound { id: u64 }
+
+#[derive(Debug, thiserror::Error, ViewError)]
+#[error("unauthorized")]
+#[view_error(status = 401)]
+struct Unauthorized;
+
+#[derive(Debug, thiserror::Error, ViewError)]
+enum EndpointError {
+    NotFound(#[from] #[view_error] InfraNotFound),
+
+    Unauthorized(#[from] #[view_error] Unauthorized)
+
+    #[error("oh no")]
+    #[view_error(user)]
+    Error1,
+}
+```
+
+## Full `derive(ViewError)` spec
+
+The macro supports enums, named structs and tuple structs (fixme correct naming).
+
+````rust
+#[derive(thiserror::Error)] // not an EditoastError
+#[error("wrong string: {0}")]
+struct WrongString(String);
+
+// error_type = "InvalidInt"
+// context = { value: number }
+#[derive(ViewError, thiserror::Error)]
+#[error("wrong int: {value}")]
+#[view_error(user, name = "InvalidInt")]
+struct WrongInt { value: i64 };
+
+#[derive(ViewError, thiserror::Error)]
+#[error("my error type")]
+#[view_error(context)]
+enum MyError {
+    // error_type = "MyError::InvalidString"
+    // context = { expected_format: string }
+    // #[view_error(internal)] by default
+    InvalidString { #[from] source: WrongString, expected_format: String },
+
+    // error_type = "MyError::InvalidInt"
+    // context = { value: number } (xyz is skipped as we forward the ViewError)
+    WrongInt { #[from] #[view_error] source: WrongInt, xyz: String }
+
+    // error_type = "MyError::Bad"
+    // context = { "0": string, "1": number }
+    #[error("user did a bad with {0} and {1}")]
+    #[view_error(user, name = "Bad")]
+    Oops(String, i64)
+}
+````
+
+### Providing `context`
+
+Only `4xx` errors require context since these are the errors the frontend may use to provide the user some form of recovery. Context is computed just before the error is serialized in `axum`'s error handler.
```

Comment:
Disagreed, in debug the context may be very valuable for 500 errors too.

Reply:
Yep. That should have been removed since the `#### Incident reports` part has been removed.

Reply:
Done in [db47135](https://github.com/OpenRailAssociation/osrd-website/pull/237/commits/db47135b8b1222ae4bf860bfa945389f4bc134fa)